### ISSUE TITLE
index 6 out of bounds for type 'int[6]'

### ISSUE THREAD

**Issue by sk0kanik** [lucasart/Demolito#170]
Clang's undefined behaviour sanitizer gives array index out of bound runtime error on position.c line 564.
https://github.com/lucasart/Demolito/blob/67072816a4f11faa2bfc2e569290b003340443ee/src/position.c#L564
I am using Android and clang is from [termux project](https://github.com/termux/termux-packages) but I think it'll be reproducible on other systems too.

```
 $ clang --version
clang version 16.0.6
Target: armv7a-unknown-linux-android24
Thread model: posix
InstalledDir: /data/data/com.termux/files/usr/bin
 $ git clone https://github.com/lucasart/Demolito
Cloning into 'Demolito'...
remote: Enumerating objects: 4245, done.
...
Resolving deltas: 100% (3005/3005), done.
 $ cd Demolito/src/
 $ nano makefile # adding -fsanitize=address -fsanitize=undefined -O1 -g
 $ make
clang -march=native -std=gnu11 -DNDEBUG -fsanitize=address -fsanitize=undefined -O1 -g -Wfatal-errors -Wall -Wextra -Wshadow -Wstrict-prototypes -Wsign-conversion -Wmissing-prototypes -Wcast-align -Wmissing-variable-declarations -Wshorten-64-to-32 -Wimplicit-int-conversion -Wimplicit-int-float-conversion -DVERSION=\"dev\" ./*.c -o demolito -s -lm -lpthread
./htable.c:53:32: warning: implicit conversion loses integer precision: 'uint64_t' (aka 'unsigned long long') to 'size_t' (aka 'unsigned int') [-Wshorten-64-to-32]
    HashCount = (hashMB << 20) / sizeof(HashEntry);
              ~ ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
1 warning generated.
./main.c:71:32: warning: implicit conversion loses integer precision: 'unsigned long long' to 'size_t' (aka 'unsigned int') [-Wshorten-64-to-32]
                uciHash = 1ULL << bb_msb((uint64_t)atoll(argv[4])); // must be a power of 2
                        ~ ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
./uci.c:79:24: warning: implicit conversion loses integer precision: 'unsigned long long' to 'size_t' (aka 'unsigned int') [-Wshorten-64-to-32]
        uciHash = 1ULL << bb_msb(uciHash); // must be a power of two
                ~ ~~~~~^~~~~~~~~~~~~~~~~~
1 warning generated.
 $ ./demolito bench
br1kq1r1/pppppp1p/3nn1p1/8/6P1/1P1NN3/P1PPPP1P/R2KQ1RB b Kkq - 0 5
position.c:564:15: runtime error: index 6 out of bounds for type 'int[6]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior position.c:564:15 in
=================================================================
==18643==ERROR: AddressSanitizer: global-buffer-overflow on address 0x039cec38 at pc 0x039ab5dc bp 0xa72fb298 sp 0xa72fb294
READ of size 4 at 0x039cec38 thread T1
    #0 0x39ab5d8  (/data/data/com.termux/files/home/sandbox/Demolito/src/demolito+0x2d5d8)
    #1 0x39b7a28  (/data/data/com.termux/files/home/sandbox/Demolito/src/demolito+0x39a28)
    #2 0x39b09d0  (/data/data/com.termux/files/home/sandbox/Demolito/src/demolito+0x329d0)
    #3 0x39aee1c  (/data/data/com.termux/files/home/sandbox/Demolito/src/demolito+0x30e1c)
    #4 0xa8e55cf0 in __pthread_start(void*) pthread_create.cpp
    #5 0xa8e0f060 in __start_thread clone.cpp

0x039cec38 is located 40 bytes before global variable 'KnightPstSeed' defined in './tune.c' (0x39cec60) of size 96
0x039cec38 is located 0 bytes after global variable 'PieceValue' defined in './tune.c' (0x39cec20) of size 24
SUMMARY: AddressSanitizer: global-buffer-overflow (/data/data/com.termux/files/home/sandbox/Demolito/src/demolito+0x2d5d8)
Shadow bytes around the buggy address:
  0x039ce980: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x039cea00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x039cea80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x039ceb00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x039ceb80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x039cec00: 00 00 00 00 00 00 00[f9]f9 f9 f9 f9 00 00 00 00
  0x039cec80: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 00
  0x039ced00: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 00
  0x039ced80: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 00
  0x039cee00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x039cee80: 00 00 00 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
Thread T1 created by T0 here:
LLVMSymbolizer: error reading file: No such file or directory
    #0 0xa8f156e4 in __interceptor_pthread_create (/data/data/com.termux/files/usr/lib/clang/16/lib/linux/libclang_rt.asan-arm-android.so+0x896e4)
    #1 0xa7a167fc  ([anon:LargeMmapAllocator]+0x2037fc)
    #2 0xa8e6501c in stdin (/apex/com.android.runtime/lib/bionic/libc.so+0xba01c) (BuildId: 58a173c6b457a368ea5ab855aaa7a660)

==18643==ABORTING
Aborted
$
```

**Comment by romanovj** [comment]
no problems on aarch64

**Comment by romanovj** [comment]
aarch64
```

~/Demolito/src $ make
clang -march=native -std=gnu11 -DNDEBUG  -fsanitize=address -fsanitize=undefined -O1 -g -flto -Wfatal-errors -Wall -Wextra -Wshadow -Wstrict-prototypes -Wsign-conversion -Wmissing-prototypes -Wcast-align -Wmissing-variable-declarations -Wshorten-64-to-32 -Wimplicit-int-conversion -Wimplicit-int-float-conversion -DVERSION=\"dev\" ./*.c -o demolito -s -lm -lpthread
~/Demolito/src $ ./demolito bench
br1kq1r1/pppppp1p/3nn1p1/8/6P1/1P1NN3/P1PPPP1P/R2KQ1RB b Kkq - 0 5
position.c:564:15: runtime error: index 6 out of bounds for type 'int[6]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior position.c:564:15 in
=================================================================
==24906==ERROR: AddressSanitizer: global-buffer-overflow on address 0x005b19c2f6f8 at pc 0x005b19c08dc0 bp 0x005df0bfab30 sp 0x005df0bfab28
READ of size 4 at 0x005b19c2f6f8 thread T1
    #0 0x5b19c08dbc  (/data/data/com.termux/files/home/Demolito/src/demolito+0x3bdbc)
    #1 0x5b19c13450  (/data/data/com.termux/files/home/Demolito/src/demolito+0x46450)
    #2 0x5b19c0d180  (/data/data/com.termux/files/home/Demolito/src/demolito+0x40180)
    #3 0x5b19c0b998  (/data/data/com.termux/files/home/Demolito/src/demolito+0x3e998)
    #4 0x7df1dc6760 in __pthread_start(void*) pthread_create.cpp
    #5 0x7df1d63470 in __start_thread clone.cpp

0x005b19c2f6f8 is located 0 bytes to the right of global variable 'PieceValue' defined in './tune.c' (0x5b19c2f6e0) of size 24
SUMMARY: AddressSanitizer: global-buffer-overflow (/data/data/com.termux/files/home/Demolito/src/demolito+0x3bdbc)
Shadow bytes around the buggy address:
  0x001b63385e80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x001b63385e90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x001b63385ea0: 00 00 00 00 00 00 00 00 04 f9 f9 f9 00 00 00 00
  0x001b63385eb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x001b63385ec0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x001b63385ed0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00[f9]
  0x001b63385ee0: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
  0x001b63385ef0: 00 00 00 00 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x001b63385f00: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x001b63385f10: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x001b63385f20: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
Thread T1 created by T0 here:
    #0 0x7df3106aa8 in pthread_create out/lib/compiler-rt-aarch64/out/llvm-project/compiler-rt/lib/asan/asan_interceptors.cpp:208:3
    #1 0x5b19c0b018  (/data/data/com.termux/files/home/Demolito/src/demolito+0x3e018)
    #2 0x5b19c00954  (/data/data/com.termux/files/home/Demolito/src/demolito+0x33954)
    #3 0x7df1d5a1f4 in __libc_init (/apex/com.android.runtime/lib64/bionic/libc.so+0x4a1f4) (BuildId: 8f3a8e9a34c28768dc32f5d6eb26ddf3)
    #4 0x5b19bf22ec  (/data/data/com.termux/files/home/Demolito/src/demolito+0x252ec)

==24906==ABORTING
Aborted
```

**Comment by lucasart** [comment]
Fixed bf19a1e
